src/VisualStudio/Core/Def/Packaging/PackageInstallerServiceFactory.cs

```diff
@@ -0,0 +1,256 @@
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.Editor.Shared.Utilities;
+using Microsoft.CodeAnalysis.Host;
+using Microsoft.CodeAnalysis.Host.Mef;
+using Microsoft.CodeAnalysis.Packaging;
+using Microsoft.VisualStudio.ComponentModelHost;
+using Microsoft.VisualStudio.LanguageServices.Implementation.ProjectSystem;
+using Microsoft.VisualStudio.Shell.Interop;
+using NuGet.VisualStudio;
+using Roslyn.Utilities;
+
+namespace Microsoft.VisualStudio.LanguageServices.Packaging
+{
+    [ExportWorkspaceServiceFactory(typeof(IPackageInstallerService)), Shared]
+    internal class PackageInstallerServiceFactory : IWorkspaceServiceFactory
+    {
+        public IWorkspaceService CreateService(HostWorkspaceServices workspaceServices)
+        {
+            return new PackageInstallerService();
+        }
+    }
+
+    internal class PackageInstallerService : ForegroundThreadAffinitizedObject, IPackageInstallerService
+    {
+        private readonly object _gate = new object();
+
+        private VisualStudioWorkspaceImpl _workspace;
+        private IVsPackageInstallerServices _packageInstallerServices;
+        private IVsPackageInstaller _packageInstaller;
+
+        private CancellationTokenSource _tokenSource = new CancellationTokenSource();
+        private bool _solutionChanged;
+        private HashSet<ProjectId> _changedProjects = new HashSet<ProjectId>();
+
+        private readonly ConcurrentDictionary<ProjectId, Dictionary<string, string>> _projectToInstalledPackageAndVersion =
+            new ConcurrentDictionary<ProjectId, Dictionary<string, string>>();
+
+        public bool TryInstallPackage(Workspace workspace, ProjectId projectId, string packageName, string versionOpt)
+        {
+            this.AssertIsForeground();
+
+            if (workspace == _workspace && _workspace != null && _packageInstallerServices != null)
+            {
+                var dte = _workspace.GetVsService<SDTE, EnvDTE.DTE>();
+                var dteProject = _workspace.TryGetDTEProject(projectId);
+                if (dteProject != null)
+                {
+                    try
+                    {
+                        if (!_packageInstallerServices.IsPackageInstalled(dteProject, packageName))
+                        {
+                            dte.StatusBar.Text = string.Format(ServicesVSResources.Installing_package_0, packageName);
+                            _packageInstaller.InstallPackage(source: null, project: dteProject, packageId: packageName, version: versionOpt, ignoreDependencies: false);
+                            dte.StatusBar.Text = string.Format(ServicesVSResources.Installing_package_0_completed, packageName);
+                        }
+
+                        return true;
+                    }
+                    catch
+                    {
+                        dte.StatusBar.Text = string.Format(ServicesVSResources.Installing_package_0_failed, packageName);
```

Comment:
I don't love the use of the StatusBar here (or above really, but especially not for failure messages). Can we tell what went wrong? Or do we have to resort to just showing the exception message? Can we escalate to a dialog or something other than the status bar?